src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,361 @@
+# TODO:
+# * add transform and transfom! functions
+# * add `Col` wrapper for whole column operations
+# * update documentation
+# * add tests
+
+# normalize_selection function makes sure that whatever input format of idx is it
+# will end up in one of four canonical forms
+# 1) Int
+# 2) AbstractVector{Int}
+# 3) Pair{Int, Pair{ColRename, Symbol}}
+# 3) Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+# 4) Pair{AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+# in this way we can easily later decide on the codepath using type signatures
+
+```
+    ColRename
+
+A singleton type indicating that column renaming operation was requested in `select`.
+```
+struct ColRename end
+
+normalize_selection(idx::AbstractIndex, sel) = idx[sel]
+normalize_selection(idx::AbstractIndex, sel::Pair{<:Any,<:Pair{<:Base.Callable,Symbol}}) =
+    idx[first(sel)] => last(sel)
+normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Symbol}) =
+    idx[first(sel)] => ColRename() => last(sel)
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    newcol = Symbol(_names(idx)[c], "_", funname(fun))
+    return c => fun => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    if length(c) > 2
+        newcol = Symbol(join(_names(idx)[c[1:2]], "_"), "_⋯_", funname(fun))
+    else
+        newcol = Symbol(join(_names(idx)[c], "_"), "_", funname(fun))
+    end
+    return c => fun => newcol
+end
+
+struct TypeHolder{T} end
+
+function select_transform_helper(th::TypeHolder{T}, cols, fun, n) where T
+    fun_transform(fun, x::T) = fun(x)
+    map(i -> fun_transform(fun, T(ntuple(j -> cols[j][i], length(cols)))), 1:n)
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Any, Any}, copycols::Bool)
+    newname = last(last(nc))
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    col_idx = first(nc)
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(last(nc)).(df[!, col_idx])
```

Comment:
It might be a bit easier to read if you stored `last(nc)` in a variable to parallel `col_idx` here and below?

Reply:
fixed